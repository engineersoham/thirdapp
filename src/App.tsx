import './App.css';
import { Route, Routes, useNavigate } from 'react-router-dom';
import RawJSON from './components/RawJSON';
import { useEffect, useState } from 'react';
import PostsTable from './components/PostsTable';

const App: React.FC = () => {
  const [page, setPage] = useState(1);
  const [storedData, setStoredData]: any = useState([]);
  const navigate = useNavigate();
  let interval: NodeJS.Timer;

  const fetchTableData = async () => {
    const response = await fetch(
      `https://hn.algolia.com/api/v1/search_by_date?query=story&page=${page}`
    );
    const resData = await response.json();
    
  console.log(resData)
    if (!storedData.includes(...resData.hits)) {
      setStoredData([...storedData,  ...resData.hits ]);
    }
  };


  const handelPage = ()=>{
    setPage((page)=>page+1)
    console.log('page called');
    // console.log(storedData)
  }

  useEffect(() => {
    fetchTableData();
  }, [page]);

  useEffect(()=>{
    interval = setInterval(handelPage,10000)
  },[])

  return (
    <div className="app">
           { <Routes>
        
        <Route
          path="/"
          element={
            <PostsTable
              storedData={storedData}
              page={page}
              setPage={setPage}
              navigate={navigate}
            />
          }
        />
         <Route path='/jsondata' element={<RawJSON/>}/>
      </Routes>}
    </div>
  );
};


export default App;