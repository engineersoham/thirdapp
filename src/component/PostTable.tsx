import { CircularProgress, Table, TableBody, TableCell, TableHead, TableRow, Typography } from '@mui/material';
import React, { useEffect } from 'react';
import { useNavigate } from 'react-router-dom';

type Props = {
    data:any
    setPage:any
}

const PostTable:React.FC<Props> = ({...props}) => {
    const navigate = useNavigate();

    const handelClick = (item:any)=>{
        alert(JSON.stringify(item))
        navigate('/rawjson',{state:item})
    }

    window.addEventListener('scroll',()=>{
        if(window.scrollY + window.innerHeight >= document.documentElement.scrollHeight){
            props.setPage((prev:any)=>prev+1)
            console.log('scroll called')
        }
    })

    useEffect(()=>{
        console.log(props.data);
    },[props.data])

  return (
    <div>
        {props.data === undefined ? (<><CircularProgress/>Loading...</> ):

        (<Table title='table'>
        <TableHead>
            <TableCell sx={{backgroundColor:'lightyellow'}}>
                <Typography variant='h6'>TITLE</Typography>
            </TableCell>
            <TableCell sx={{backgroundColor:'lightyellow'}} >
                <Typography variant='h6'>URL</Typography>
            </TableCell>
            <TableCell sx={{backgroundColor:'lightyellow'}}>
                <Typography variant='h6'>CREATED_AT</Typography>
            </TableCell>
            <TableCell sx={{backgroundColor:'lightyellow'}}>
                <Typography variant='h6'>AUTHOR</Typography>
            </TableCell>
        </TableHead>
        <TableBody>
            {Array.from(props.data).map((item:any,idx:any)=>{
                return(
                    <TableRow key={idx} onClick={()=>handelClick(item)}>
                        <TableCell>
                            {item.title ? item.title : <i>data not found</i> }
                        </TableCell>
                        <TableCell>
                            {item.url ? item.url : <i>data not found</i> }
                        </TableCell>
                        <TableCell>
                            {item.created_at ? item.created_at : <i>data not found</i> }
                        </TableCell>
                        <TableCell>
                            {item.author ? item.author : <i>data not found</i> }
                        </TableCell>
                    </TableRow>
                )
            })}
        </TableBody>
    </Table>)
        }
    </div>
  )
}

export default PostTable