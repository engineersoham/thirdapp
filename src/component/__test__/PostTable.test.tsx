import { screen, render, fireEvent } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";
import App from "../../App";
import PostTable from "../PostTable"

it('snap of table',()=>{
    const tree = render(
        <BrowserRouter>
        <PostTable data={()=>{}} setPage={()=>{}}/>
        </BrowserRouter>
    )
    expect(tree).toMatchSnapshot()
    screen.debug()
    expect(screen.getByTitle('table')).toBeInTheDocument()
})

describe('post table fetch', ()=>{
    jest.setTimeout(20000);
    it("fetch test",async()=>{
        render(
             <BrowserRouter>
            <PostTable data={[{created_at: "11", author:"aa"}]} 
            setPage={()=>{}}/>
            </BrowserRouter>
        );
        await new Promise((r)=>setTimeout(r, 5000));
        const post = screen.getByTestId("post-0")
        expect(post).toBeInTheDocument()
        const theClick = fireEvent.click(post);
        expect(theClick).toBeTruthy();
        fireEvent.scroll(window, {target:{scrollY:100}});
    })
})

