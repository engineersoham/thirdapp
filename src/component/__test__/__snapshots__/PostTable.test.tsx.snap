// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snap of table 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <table
          class="MuiTable-root css-rqglhn-MuiTable-root"
          title="table"
        >
          <thead
            class="MuiTableHead-root css-15wwp11-MuiTableHead-root"
          >
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ei063q-MuiTableCell-root"
              scope="col"
            >
              <h6
                class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
              >
                TITLE
              </h6>
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ei063q-MuiTableCell-root"
              scope="col"
            >
              <h6
                class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
              >
                URL
              </h6>
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ei063q-MuiTableCell-root"
              scope="col"
            >
              <h6
                class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
              >
                CREATED_AT
              </h6>
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ei063q-MuiTableCell-root"
              scope="col"
            >
              <h6
                class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
              >
                AUTHOR
              </h6>
            </th>
          </thead>
          <tbody
            class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
          />
        </table>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <table
        class="MuiTable-root css-rqglhn-MuiTable-root"
        title="table"
      >
        <thead
          class="MuiTableHead-root css-15wwp11-MuiTableHead-root"
        >
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ei063q-MuiTableCell-root"
            scope="col"
          >
            <h6
              class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
            >
              TITLE
            </h6>
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ei063q-MuiTableCell-root"
            scope="col"
          >
            <h6
              class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
            >
              URL
            </h6>
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ei063q-MuiTableCell-root"
            scope="col"
          >
            <h6
              class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
            >
              CREATED_AT
            </h6>
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ei063q-MuiTableCell-root"
            scope="col"
          >
            <h6
              class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
            >
              AUTHOR
            </h6>
          </th>
        </thead>
        <tbody
          class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
        />
      </table>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
